server/src/main/java/org/opensearch/search/optimization/filterrewrite/OptimizationContext.java

```diff
@@ -0,0 +1,188 @@
+
+import static org.opensearch.search.optimization.filterrewrite.Helper.loggerName;
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+/**
+ * Context object for doing the filter rewrite optimization in ranges type aggregation
+ * <p>
+ * This holds the common business logic and delegate aggregator-specific logic to {@link AggregatorBridge}
+ *
+ * @opensearch.internal
+ */
+public final class OptimizationContext {
+
+    private static final Logger logger = LogManager.getLogger(loggerName);
+
+    private boolean canOptimize = false;
+    private boolean preparedAtShardLevel = false;
+
+    final AggregatorBridge aggregatorBridge;
+    int maxAggRewriteFilters;
+    String shardId;
+
+    private Ranges ranges;
+    private Ranges rangesFromSegment;
+
+    // debug info related fields
+    private int leaf;
+    private int inner;
+    private int segments;
+    private int optimizedSegments;
+
+    public OptimizationContext(AggregatorBridge aggregatorBridge) {
+        this.aggregatorBridge = aggregatorBridge;
+    }
+
+    public boolean canOptimize(final Object parent, final int subAggLength, SearchContext context) {
+        if (context.maxAggRewriteFilters() == 0) return false;
+
+        if (parent != null || subAggLength != 0) return false;
+
+        this.canOptimize = aggregatorBridge.canOptimize();
+        if (canOptimize) {
+            aggregatorBridge.setOptimizationContext(this);
+            this.maxAggRewriteFilters = context.maxAggRewriteFilters();
+            this.shardId = context.indexShard().shardId().toString();
+        }
+        logger.debug("Fast filter rewriteable: {} for shard {}", canOptimize, shardId);
+        return canOptimize;
+    }
+
+    public void prepare() throws IOException {
+        assert ranges == null : "Ranges should only be built once at shard level, but they are already built";
+        aggregatorBridge.prepare();
+        if (ranges != null) {
+            preparedAtShardLevel = true;
+        }
+    }
+
+    public void prepareFromSegment(LeafReaderContext leaf) throws IOException {
+        aggregatorBridge.prepareFromSegment(leaf);
+    }
+
+    void setRanges(Ranges ranges) {
+        this.ranges = ranges;
+    }
+
+    void setRangesFromSegment(Ranges ranges) {
+        this.rangesFromSegment = ranges;
+    }
+
+    Ranges getRanges() {
+        if (rangesFromSegment != null) return rangesFromSegment;
+        return ranges;
+    }
+
+    /**
+     * Try to populate the bucket doc counts for aggregation
+     * <p>
+     * Usage: invoked at segment level — in getLeafCollector of aggregator
+     *
+     * @param incrementDocCount consume the doc_count results for certain ordinal
+     * @param segmentMatchAll if your optimization can prepareFromSegment, you should pass in this flag to decide whether to prepareFromSegment
+     */
+    public boolean tryOptimize(final LeafReaderContext leafCtx, final BiConsumer<Long, Long> incrementDocCount, boolean segmentMatchAll)
```

Comment:
This will be invoked by multiple threads with concurrent search enabled.  This doesn't appear to be thread safe. Particularly calls to `tryBuildRangesFromSegment ` and then `setRangesFromSegment` where state `    private Ranges rangesFromSegment;` in this class will be set from separate threads.

Reply:
Valid concern, and it will probably lead to strange results.

I don't think we should resolve this in this refactor PR. Create one task tracking [this](https://github.com/orgs/opensearch-project/projects/153/views/8?filterQuery=&pane=issue&itemId=73763069) 

I will see if there's a easy way to not choose the optimization when concurrent segment search enabled.

Reply:
Tried to solve the concern of `setRangesFromSegment ` in this commit, as the change is small — using a map to save ranges of different segments. I think it's ok to have this along with this refactoring.

[`234eb44` (#14464)](https://github.com/opensearch-project/OpenSearch/pull/14464/commits/234eb446a59a2b23b697f98e14a674dbf17c9bac)